File: generated-code/typescript/codecraft/TcpReadWrite/src/model/repair-properties.ts
import { EntityType } from "./entity-type";
import { Stream } from "../stream";

/**
 * Entity's repair properties
 */
export class RepairProperties {
    /**
     * Valid target entity types
     */
    validTargets: Array<EntityType>
    /**
     * Health restored in one tick
     */
    power: number

    constructor(validTargets: Array<EntityType>, power: number) {
        this.validTargets = validTargets;
        this.power = power;
    }

    /**
     * Read RepairProperties from input stream
     */
    static async readFrom(stream: Stream): Promise<RepairProperties> {
        let validTargets;
        validTargets = [];
        for (let validTargetsCount = await stream.readInt(); validTargetsCount > 0; validTargetsCount--) {
            let validTargetsElement;
            validTargetsElement = await EntityType.readFrom(stream);
            validTargets.push(validTargetsElement);
        }
        let power;
        power = await stream.readInt();
        return new RepairProperties(validTargets, power)
    }

    /**
     * Write RepairProperties to output stream
     */
    async writeTo(stream: Stream) {
        let validTargets = this.validTargets;
        await stream.writeInt(validTargets.length);
        for (let validTargetsElement of validTargets) {
            await validTargetsElement.writeTo(stream);
        }
        let power = this.power;
        await stream.writeInt(power);
    }
}

File: generated-code/python/codecraft/TcpReadWrite/model/repair_properties.py
from model.entity_type import EntityType
from stream_wrapper import StreamWrapper
from typing import List

class RepairProperties:
    """Entity's repair properties"""

    __slots__ = ("valid_targets","power",)

    valid_targets: List[EntityType]
    power: int

    def __init__(self, valid_targets: List[EntityType], power: int):
        self.valid_targets = valid_targets
        """Valid target entity types"""
        self.power = power
        """Health restored in one tick"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "RepairProperties":
        """Read RepairProperties from input stream
        """
        valid_targets = []
        for _ in range(stream.read_int()):
            valid_targets_element = EntityType(stream.read_int())
            valid_targets.append(valid_targets_element)
        power = stream.read_int()
        return RepairProperties(valid_targets, power)
    
    def write_to(self, stream: StreamWrapper):
        """Write RepairProperties to output stream
        """
        stream.write_int(len(self.valid_targets))
        for element in self.valid_targets:
            stream.write_int(element)
        stream.write_int(self.power)
    
    def __repr__(self):
        return "RepairProperties(" + \
            repr(self.valid_targets) + \
            ", " + \
            repr(self.power) + \
            ")"